File: 1695-maximum-erasure-value/1695-maximum-erasure-value.py
class Solution:
    def maximumUniqueSubarray(self, nums: List[int]) -> int:
        begin, end, Set, summ = 0, 0, set(), 0
        ans = 0
        
        # 2-pointer or sliding window approach 
        while end < len(nums):
            if nums[end] not in Set:
                summ += nums[end]
                Set.add(nums[end])
                end += 1
                ans = max(ans, summ)
            else:
                summ -= nums[begin]
                Set.remove(nums[begin])
                begin += 1
            
        return ans
        

File: 1695-maximum-erasure-value/1695-maximum-erasure-value.cpp
class Solution {
public:
    int maximumUniqueSubarray(vector<int>& nums) {
        int result = 0;
        unordered_set<int> hset;
        
        for(int i = 0, j = 0, window = 0; j < nums.size(); j++){
            while (hset.find(nums[j]) != hset.end()){
                hset.erase(nums[i]);
                window -= nums[i];
                i++;
            }
            hset.insert(nums[j]);
            window += nums[j];
            result = max(result, window);
        }
        return result;
        
    }
};